### PR TITLE
Reduce compaction serialization

### PR DESCRIPTION
**Pull request by mikolajsieluzycki**
update_history can take a long time compared to compaction, as a call
issued on shard S1 can be handled on shard S2. If the other shard is
under heavy load, we may unnecessarily block kicking off a new
compaction. Normally it isn't a problem, as compactions aren't super
frequent, but there were edge cases where the described behaviour caused
compaction to fail to keep up with excessive flushing, leading to too
many sstables on disk and OOM during a read.

There is no need to wait with next compaction until history is updated,
so release the weight earlier to remove unnecessary serialization.

Changelog:
v3:
- explicitly call deregister instead of moving the weight RAII object to release weight
- mark compaction as finished when sstables are compacted, without waiting for history to update

v2:
- Split the patches differently for easier review
- Rebased agains newer master, which contains fixes that failed the debug version of the test
- Removed the test, as it will be provided by [PR#10717](https://github.com/scylladb/scylla/pull/10717)

### CODE REVIEW

**Comment by avikivity** [comment]
> Compaction weight is used to serialize similar compactions. Updating compaction history table can potentially take long time compared to the compaction itself, and there is no need for it to serialize compactions. Releasing weight before history is updated increases compaction throughput and reduces chances of having a lot of small sstables if compaction doesn't keep up with flushing.


Is it really a problem? The compaction history table update should be just adding a memtable entry.

**Comment by mikolajsieluzycki** [comment]
@avikivity Most of the time - no. In specific scenarios (investigating what those are) it takes 300ms, and if you have flush spam then it may become a problem.

**Comment by avikivity** [comment]
Why does adding a memtable entry take 300ms?

**Comment by avikivity** [comment]
Maybe it's the commitlog?

**Comment by raphaelsc** [comment]
I don't like the fact that compaction history update isn't being handled locally. Perhaps we can make compaction UUID (used as pkey for history table) shard aware, so we don't flood smp queues with those requests. Can affect scalability, right? There will always be small sized jobs, which tend to complete frequently. It's a waste of resources to have each shard possibly issuing a cross-cpu request for every small-sized compaction completed.

@avikivity what do you think?

**Comment by avikivity** [comment]
> I don't like the fact that compaction history update isn't being handled locally. Perhaps we can make compaction UUID (used as pkey for history table) shard aware, so we don't flood smp queues with those requests. Can affect scalability, right? There will always be small sized jobs, which tend to complete frequently. It's a waste of resources to have each shard possibly issuing a cross-cpu request for every compaction completed.
> 
> @avikivity what do you think?

It's rare enough that we can avoid adding extra complexity for it. Compactions are rarer than once every second, which is perfectly okay for cross-shard ops.

**Comment by bhalevy** [comment]
@raphaelsc / @mikolajsieluzycki anything blocking this PR?

@mikolajsieluzycki please rebase, retest, and start a new CI job for it.

**Comment by mikolajsieluzycki** [comment]
@bhalevy The failed test in debug has been fixed, I'll rebase and resend

File: compaction/compaction_manager.cc
Comment (denesb) [inline compaction/compaction_manager.cc:903]: Can't we do it like: `weight_r = {};`?

**Review by denesb — APPROVED** [review]
LGTM, one nitpick.
I feel like the refactoring could have also been done with much less steps (commits), but it is fine as-is.
I need an ACK from @avikivity before merging, as he had questions and it is not clear he was happy with the answers.

File: compaction/compaction_manager.cc
Comment (denesb) [reply compaction/compaction_manager.cc:903]: Please do. This is a really minor thing, but it does expresses intent much better.

File: compaction/compaction_manager.cc
Comment (avikivity) [inline compaction/compaction_manager.cc:903]: nit: this should be outside the conditional. Otherwise, after the conditional, the state of weight registration is unspecified.

Comment (avikivity) [reply compaction/compaction_manager.cc:903]: There's almost nothing after the conditional, but one day, there may be. So better to be prepared.

File: compaction/compaction_manager.cc
Comment (mikolajsieluzycki) [reply compaction/compaction_manager.cc:888]: Fair point. @raphaelsc is it safe to call maybe_retry if compaction succeeded, but update_history failed? If so, I could wrap the update_history in another try/catch block to handle its exception slightly differently and call finish_compaction(state::failed) only if the exception came from compact_sstables. Something like:

```
            std::exception_ptr ex;

            try {
                bool should_update_history = this->should_update_history(descriptor.options.type());
                sstables::compaction_result res = co_await compact_sstables(std::move(descriptor), _compaction_data, std::move(release_exhausted));
                finish_compaction();
                weight_r.deregister();
                if (should_update_history) {
                    try {
                        co_await update_history(*_compacting_table, res, _compaction_data);
                    } catch (...) {
                        ex = std::current_exception();
                    }
                }
                if (!ex) {
                    _cm.reevaluate_postponed_compactions();
                    continue;
                }
            } catch (...) {
                finish_compaction(state::failed);
                ex = std::current_exception();
            }

            if ((co_await maybe_retry(std::move(ex))) == stop_iteration::yes) {
                co_return;
            }
```

Comment (avikivity) [reply compaction/compaction_manager.cc:888]: No need to do it twice.

File: compaction/compaction_manager.cc
Comment (mikolajsieluzycki) [reply compaction/compaction_manager.cc]: Fixed in v4

**Comment by mikolajsieluzycki** [comment]
v4:
- Moved weight deregistration before the conditional history update

**Comment by avikivity** [comment]
We put backport candidate labels on issues, not pull requests.

**Pull request closed without merging**
